Sources/xcodeproj/Utils/FileWriter.swift

```diff
@@ -0,0 +1,81 @@
+import Foundation
+import PathKit
+
+/// This protocol defines the interface to write project files to disk.
+protocol FileWriting {
+    
+    /// Gracefully writes project data to disk:
+    ///  1. If the content hasn't changed, it doesn't override the existing file.
+    ///  2. If the content has changed, it overrides the existing file.
+    ///
+    /// - Parameters:
+    ///   - data: Data to be writen.
+    ///   - to: Path where the data will be written to.
+    /// - Throws: An error if the write fails
+    /// - Returns: True if the file was writen.
+    @discardableResult
+    func write(data: Data, to: Path) throws -> Bool
+    
+    /// Gracefully writes project strings to disk:
+    ///  1. If the content hasn't changed, it doesn't override the existing file.
+    ///  2. If the content has changed, it overrides the existing file.
+    ///
+    /// - Parameters:
+    ///   - string: String to be writen.
+    ///   - to: Path where the data will be written to.
+    /// - Throws: An error if the write fails
+    /// - Returns: True if the file was writen.
+    @discardableResult
+    func write(string: String, to: Path) throws -> Bool
+}
+
+class FileWriter: FileWriting {
+    /// Gracefully writes project data to disk:
+    ///  1. If the content hasn't changed, it doesn't override the existing file.
+    ///  2. If the content has changed, it overrides the existing file.
+    ///
+    /// - Parameters:
+    ///   - data: Data to be writen.
+    ///   - to: Path where the data will be written to.
+    /// - Throws: An error if the write fails
+    /// - Returns: True if the file was writen.
+    @discardableResult
+    func write(data: Data, to: Path) throws -> Bool {
+        var existingData: Data?
+        if to.exists {
```

Comment:
I'm not 100% sure about this one because we are loading the file into memory to do the comparison. I wonder how negative would be the impact on the memory footprint. Is there any chance you can try it with your project at Lyft @keith ?

The alternative that I considered for this one was writing the data to a temporary directory and compare the checksum of the two files. For that I'd need o some shell abstraction *(compatible with Linux)* to run the `md5` command. What do you think @yonaskolb?

Reply:
Yea in our case we wouldn't want this since we implement our own comparison checking and only regenerate the project if we need to in the first place. So whenever we run XcodeGen we know it needs to write it. We do this by diffing git ls-files and our project spec.

Reply:
What about making this opt-in feature? @yonaskolb XcodeGen could have support for a flag for changing the writing strategy.